File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        prefix = [1]
        postfix = [1]
        output = []
        mult1 = 1
        mult2 = 1
        i = 1
        j  = len(nums) - 2
        while(i  < len(nums) and j >= 0):
            mult1 *= nums[i-1]
            prefix.append(mult1)
            i += 1

            mult2 *= nums[j + 1]
            postfix.append(mult2)
            j -= 1
        postfix = postfix[::-1]
        i, j  = 0, 0
        while i < len(nums):
            output.append(postfix[i] * prefix[i])
            i+= 1
        return output

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int prefix[] = new int[n];
        int postfix[] = new int[n];

        prefix[0] = 1;
        postfix[n - 1] = 1;
        int i = 1, j = n - 2;
        int m1 = 1, m2 = 1;
        while(i < n && j >= 0){
            m1 *= nums[i - 1];
            m2 *= nums[j + 1];
            postfix[j] = m2;
            prefix[i] = m1;
            i++;
            j--;
        }
        
        i = 0;
        int result[] = new int[n], k = 0;
        while(i < n ) {
            result[k++] = prefix[i] * postfix[i];
            i++;
        }
        return result;
    }
}